src/lib.rs

```diff
@@ -65,27 +70,88 @@ impl Otel {
+                }
+            }
+        }
+        if let Some(log_targets) = config.log_export_targets {
+            for log_target in log_targets {
+                if let Some(ca_cert_path) = &log_target.ca_cert_path {
+                    ca_cert_paths.insert(ca_cert_path.clone());
+                }
+            }
+        }
+
+        let (shutdown_tx, shutdown_rx) = mpsc::channel(1);
+
         Otel {
             registry,
             meter_provider,
             logger_provider,
+            ca_cert_paths,
+            shutdown_tx,
+            shutdown_rx,
         }
     }
 
     /// Long running tasks for otel propagation.
-    pub async fn run(&self) {
-        if let Some(prometheus_registry) = &self.registry {
-            let _ = httpserver_init(
-                prometheus_registry.port,
-                prometheus_registry.registry.clone(),
-            )
-            .await;
+    ///
+    /// # Errors
+    /// * `OtelError::WatcherError` - If there is an error with the CA file watcher
+    pub async fn run(&mut self) -> Result<(), OtelError> {
+        // Create watcher task to detect changes to the CA files
+        let (mut ca_watcher, mut ca_watcher_rx) = create_watcher()?;
+        self.ca_cert_paths.iter().for_each(|path| {
+            let _ = ca_watcher
+                .watcher()
+                .watch(&PathBuf::from(path), RecursiveMode::NonRecursive);
+            debug!("watching for changes to {path}");
+        });
+
+        tokio::select! {
+            notification = ca_watcher_rx.recv(), if !self.ca_cert_paths.is_empty() => {
+                match notification {
+                    Some(Ok(event)) => {
+                        info!("CA cert path changed: {event:?}");
+                    }
+                    Some(Err(e)) => {
+                        error!("error received from CA file watcher: {e}");
+                    }
+                    None => {
+                        warn!("CA file watcher channel closed.");
+                    }
+                }
+                info!("Exiting to allow restart of component.");
+                return Err(OtelError::CaWatcherEvent)
+            }
+            () = async {
+                if let Some(prometheus_registry) = &self.registry {
+                    let _ = httpserver_init(
+                        prometheus_registry.port,
+                        prometheus_registry.registry.clone(),
+                    ).await;
+                }
+            }, if self.registry.is_some() => {
+                info!("prometheus server stopped. Exiting.");
+                return Err(OtelError::PrometheusServerStopped)
+            }
+            _ = self.shutdown_rx.recv() => {
+                info!("shutting down otel component");
+            }
         }
+
+        Ok(())
     }
 
     /// Graceful shutdown that flushes any pending metrics and logs to the exporter.
-    pub fn shutdown(&self) {
+    pub async fn shutdown(&self) {
```

Comment:
Just a question, is there a particular reason for changing shutdown() to async? I wondered because I think having shutdown as a synchronous function makes more sense for its purpose.